pennylane_lightning/src/algorithms/AdjointDiff.hpp

```diff
@@ -490,6 +513,100 @@ template <class T = double> class AdjointJacobian {
+                                 const StateVectorManaged<T> &reference_state,
+                                 const std::vector<ObsDatum<T>> &observables) {
+        // Globally scoped exception value to be captured within OpenMP block.
+        // See the following for OpenMP design decisions:
+        // https://www.openmp.org/wp-content/uploads/openmp-examples-4.5.0.pdf
+        std::exception_ptr ex = nullptr;
+        size_t num_observables = observables.size();
+#if defined _OPENMP
+#pragma omp parallel default(none)                                             \
+    shared(states, reference_state, observables, ex, num_observables)
+        {
+#pragma omp for
+#endif
+            for (size_t h_i = 0; h_i < num_observables; h_i++) {
+                try {
+                    states[h_i].updateData(reference_state.getDataVector());
+                    applyObservable(states[h_i], observables[h_i]);
+                } catch (...) {
+#if defined _OPENMP
+#pragma omp critical
+#endif
+                    ex = std::current_exception();
+#if defined _OPENMP
+#pragma omp cancel for
+#endif
+                }
+            }
+#if defined _OPENMP
+            if (ex) {
+#pragma omp cancel parallel
+            }
+        }
+#endif
+        if (ex) {
+            std::rethrow_exception(ex);
+        }
+    }
+
+    /**
+     * @brief OpenMP accelerated application of adjoint operations to
+     * statevectors.
+     *
+     * @param states Vector of all statevectors; 1 per observable
+     * @param operations Operations list.
+     * @param op_idx Index of given operation within operations list to take
+     * adjoint of.
+     */
+    inline void applyOperationsAdj(std::vector<StateVectorManaged<T>> &states,
+                                   const OpsData<T> &operations,
+                                   size_t op_idx) {
+        // Globally scoped exception value to be captured within OpenMP block.
+        // See the following for OpenMP design decisions:
+        // https://www.openmp.org/wp-content/uploads/openmp-examples-4.5.0.pdf
+        std::exception_ptr ex = nullptr;
+        size_t num_states = states.size();
+#if defined _OPENMP
+#pragma omp parallel default(none)                                             \
+    shared(states, operations, op_idx, ex, num_states)
+        {
+#pragma omp for
+#endif
+            for (size_t obs_idx = 0; obs_idx < num_states; obs_idx++) {
+                try {
+                    applyOperationAdj(states[obs_idx], operations, op_idx);
+                } catch (...) {
+#if defined _OPENMP
+#pragma omp critical
+#endif
+                    ex = std::current_exception();
+#if defined _OPENMP
+#pragma omp cancel for
+#endif
+                }
+            }
+#if defined _OPENMP
+            if (ex) {
+#pragma omp cancel parallel
+            }
+        }
+#endif
+        if (ex) {
+            std::rethrow_exception(ex);
+        }
+    }
+
```

Comment:
Are the above additions purely resulting from `clang-tidy` recommendations? :thinking:

Reply:
Nice!